SubscriptionHandler type issue

The `SubscriptionHandler` used for generating a composite response from subscription events is currently typed as
```typescript
export type SubscriptionHandler<T, R> = (prev: R | void, data: T) => R;
```
but the `void` type should not be used in argument types unless it is to disallow them from being used. For instance, the trivial handler
```typescript
(updates: number = 0, response) => {
      return 42
    }
```
errors out with
```
error TS2345: Argument of type '(updates: number | undefined, response: ServerImageEventsSubscription) => number' is not assignable to parameter of type 'SubscriptionHandler<ServerImageEventsSubscription, ServerImageEventsSubscription>'.
  Types of parameters 'updates' and 'prev' are incompatible.
    Type 'void | ServerImageEventsSubscription' is not assignable to type 'number | undefined'.
      Type 'void' is not assignable to type 'number | undefined'.

151     (updates: number = 0, response) => {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I think that the intention with this handler is to pass in `undefined` on the first call, when there is no previous state, and the logic of `useSubscription` indeed does this. Therefore, I think that this type should be
```typescript
export type SubscriptionHandler<T, R> = (prev: R | undefined, data: T) => R;
```

EDIT:
It seems that there is another subtle issue here: because `useSubscription` sets `R=T` by default in its type parameters, the `SubscriptionHandler` expects the "state" to be of the same type as the subscription data, unless explicitly overriden. However, using
```typescript
useSubscription<
    ServerImageEventsSubscription,
    number
  >(
    {
      query: ServerImageEventsDocument,
    },
    (updates: number = 0, response) => {
      return 4
    }
  )
```

still gives a similar (but more sane) error
```
 error TS2345: Argument of type '(updates: number | undefined, response: ServerImageEventsSubscription) => number' is not assignable to parameter of type 'SubscriptionHandler<ServerImageEventsSubscription, number>'.
  Types of parameters 'updates' and 'prev' are incompatible.
    Type 'number | void' is not assignable to type 'number | undefined'.
      Type 'void' is not assignable to type 'number | undefined'.

152     (updates: number = 0, response) => {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/docs/api.md
+++ b/docs/api.md
@@ -73,7 +73,7 @@ interface UseSubscriptionArgs {
 And the handler has the signature:
 
 ```js
-type SubscriptionHandler<T, R> = (prev: R | void, data: T) => R;
+type SubscriptionHandler<T, R> = (prev: R | undefined, data: T) => R;
 ```
 
 Meaning that the subscription handler receives the previous data or undefined
@@ -139,12 +139,12 @@ interface UseSubscriptionState<T> {
 
 #### Props
 
-| Prop      | Type                                      | Description                                                                                           |
-| --------- | ----------------------------------------- | ----------------------------------------------------------------------------------------------------- |
-| query     | `string`                                  | The GraphQL subscription's query                                                                      |
-| variables | `object`                                  | The GraphQL subscriptions' variables                                                                  |
-| handler   | `void \| (prev: R \| void, data: T) => R` | The handler that should combine/update the subscription's data with incoming data                     |
-| children  | `RenderProps => ReactNode`                | A function that follows the typical render props pattern. The shape of the render props is as follows |
+| Prop      | Type                                                | Description                                                                                           |
+| --------- | --------------------------------------------------- | ----------------------------------------------------------------------------------------------------- |
+| query     | `string`                                            | The GraphQL subscription's query                                                                      |
+| variables | `object`                                            | The GraphQL subscriptions' variables                                                                  |
+| handler   | `undefined \| (prev: R \| undefined, data: T) => R` | The handler that should combine/update the subscription's data with incoming data                     |
+| children  | `RenderProps => ReactNode`                          | A function that follows the typical render props pattern. The shape of the render props is as follows |
 
 #### Render Props
 

--- a/examples/2-using-subscriptions/README.md
+++ b/examples/2-using-subscriptions/README.md
@@ -1,21 +1,35 @@
 ## About
 
-This is a basic example of subscriptions being used with URQL.
+This is a basic example of subscriptions being used with `urql`. It demonstrates some of the core principles around wiring up subscriptions, including:
+
+- Setting up the `subscriptionExchange` with a `subscriptionClient`.
+- Using the `useSubscription` hook.
+- Using a `handler` to accumulate the results of subscriptions.
 
 ## Usage
 
-#### 1. Install dependencies
+#### 1. Install dependencies in repo root.
+
+To get started with the example, make sure to install `urql`'s dependencies in the root directory of your clone. We `link` to the `urql` source from the `examples` directory, so its dependencies need to be installed.
 
 ```bash
+# In root directory
 yarn
 ```
 
-#### 2. Start server
+#### 2. Navigate to this directory and install dependencies.
+
+```bash
+cd examples/2-using-subscriptions
+yarn
+```
+
+#### 3. Start the example app and server (this is done in a single command).
 
 ```bash
 yarn start
 ```
 
-#### 3. Open browser
+#### 4. Open your browser
 
 Navigate to example at [http://localhost:3000/](http://localhost:3000/).

--- a/src/hooks/useMutation.test.tsx
+++ b/src/hooks/useMutation.test.tsx
@@ -113,7 +113,7 @@ describe('on execute', () => {
       execute(vars);
     });
     expect(client.executeMutation.mock.calls[0][1]).toHaveProperty('meta', {
-      source: 'MutationUser',
+      source: 'Object',
     });
   });
 });

--- a/src/hooks/useQuery.test.tsx
+++ b/src/hooks/useQuery.test.tsx
@@ -96,7 +96,7 @@ describe('on initial useEffect', () => {
       expect.any(Object),
       expect.objectContaining({
         meta: {
-          source: 'QueryUser',
+          source: 'Object',
         },
       })
     );

--- a/src/hooks/useRequest.ts
+++ b/src/hooks/useRequest.ts
@@ -8,7 +8,7 @@ export const useRequest = (
   query: string | DocumentNode,
   variables?: any
 ): GraphQLRequest => {
-  const prev = useRef<void | GraphQLRequest>(undefined);
+  const prev = useRef<undefined | GraphQLRequest>(undefined);
 
   return useMemo(() => {
     const request = createRequest(query, variables);

--- a/src/hooks/useSubscription.test.tsx
+++ b/src/hooks/useSubscription.test.tsx
@@ -65,7 +65,7 @@ describe('on initial useEffect', () => {
     renderer.create(<SubscriptionUser q={query} />);
     expect(client.executeSubscription).toBeCalledWith(
       expect.any(Object),
-      expect.objectContaining({ meta: { source: 'SubscriptionUser' } })
+      expect.objectContaining({ meta: { source: 'Object' } })
     );
   });
 });

--- a/src/hooks/useSubscription.ts
+++ b/src/hooks/useSubscription.ts
@@ -12,7 +12,7 @@ export interface UseSubscriptionArgs<V> {
   variables?: V;
 }
 
-export type SubscriptionHandler<T, R> = (prev: R | void, data: T) => R;
+export type SubscriptionHandler<T, R> = (prev: R | undefined, data: T) => R;
 
 export interface UseSubscriptionState<T> {
   fetching: boolean;

